discovery_engine_core/core/src/stack/filters/article.rs

```diff
@@ -39,52 +39,70 @@ impl ArticleFilter for DuplicateFilter {
-            .chain(stack.iter().map(|doc| doc.resource.url.as_str()))
-            .collect::<HashSet<_>>();
+        // discard dups in the title keeping only the best ranked
+        articles.sort_unstable_by(|art1, art2| {
+            match art1.title.to_lowercase().cmp(&art2.title.to_lowercase()) {
+                Ordering::Equal => art1.rank.cmp(&art2.rank),
+                ord => ord,
+            }
+        });
+        articles.dedup_by_key(|art| art.title.to_lowercase());
 
-        let titles = history
-            .iter()
-            .map(|doc| &doc.title)
-            .chain(stack.iter().map(|doc| &doc.resource.title))
-            .collect::<HashSet<_>>();
+        // discard dups in the link (such dups assumed to have the same rank)
+        articles.sort_unstable_by(|art1, art2| art1.link.cmp(&art2.link));
+        articles.dedup_by(|art1, art2| art1.link == art2.link);
 
-        let mut articles: BTreeSet<_> = articles.into_iter().map(UniqueArticle).collect();
+        let (hist_urls, hist_titles) = history
+            .iter()
+            .map(|doc| (doc.url.as_str(), doc.title.to_lowercase().trim().to_owned()))
+            .unzip::<_, _, HashSet<_>, HashSet<_>>();
 
-        articles.retain(|article| {
-            !(urls.contains(&article.0.link.as_str()) || titles.contains(&&article.0.title))
+        // discard dups of historical documents
+        articles.retain(|art| {
+            !hist_titles.contains(art.title.to_lowercase().trim())
+                && !hist_urls.contains(art.link.as_str())
         });
 
-        return Ok(articles.into_iter().map(|u| u.0).collect());
-
-        struct UniqueArticle(Article);
+        let stack_urls = stack
+            .iter()
+            .map(|doc| doc.resource.url.as_str())
+            .collect::<HashSet<_>>();
 
-        impl Eq for UniqueArticle {}
-        impl PartialEq for UniqueArticle {
-            fn eq(&self, other: &Self) -> bool {
-                self.0.link == other.0.link || self.0.title == other.0.title
+        let stack_titles = stack
+            .iter()
+            .map(|doc| {
+                let NewsResource { title, rank, .. } = &doc.resource;
+                (title.to_lowercase().trim().to_owned(), *rank)
+            })
+            .fold(HashMap::new(), |mut titles, (title, rank)| {
+                titles
+                    .entry(title)
+                    .and_modify(|best_rank| {
+                        if rank < *best_rank {
+                            *best_rank = rank;
+                        }
+                    })
+                    .or_insert(rank);
+                titles
+            });
+
+        // discard worse-ranked dups of stack documents; more precisely, discard:
+        // * dups of stack documents in the url
+        // * dups of stack documents in the title when the rank is no better
+        articles.retain(|art| {
+            match (
+                stack_urls.contains(art.link.as_str()),
```

Comment:
Can we avoid to look into titles if we know that it is contained in the url?
Something like
`urls.contain(...) || titles.get(...).map(|doc| art.rank < doc.rank).unwrap_or(true)`

Reply:
nice suggestion. based on your idea,
```rust
!urls.contains(art.link) && titles.get(art.title).map_or(true, |doc_rank| art.rank < doc_rank)
```
should do the trick (clippy helpfully suggested `map_or` in place of `.map().unwrap_or()`)